binja_script.py

```diff
@@ -0,0 +1,400 @@
+from __future__ import annotations
+
+import ctypes
+import sys
+import threading
+import xmlrpc.client
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Tuple
+from xmlrpc.server import SimpleXMLRPCRequestHandler
+from xmlrpc.server import SimpleXMLRPCServer
+
+import binaryninja
+from binaryninja.enums import RegisterValueType
+from binaryninja.enums import VariableSourceType
+
+# Allow large integers to be transmitted
+xmlrpc.client.MAXINT = 10**100
+xmlrpc.client.MININT = -(10**100)
+
+
+host = "127.0.0.1"
+port = 31337
+
+logger = binaryninja.log.Logger(0, "pwndbg-integration")
+
+
+class CustomLogHandler(SimpleXMLRPCRequestHandler):
+    def log_message(self, format: str, *args: Any):
+        logger.log_debug(format % args)
+
+
+# get the earliest-starting function that contains a given address
+def get_widest_func(bv: binaryninja.BinaryView, addr: int) -> binaryninja.Function | None:
+    funcs = bv.get_functions_containing(addr)
+    if len(funcs) == 0:
+        return None
+    return min(funcs, key=lambda f: f.start)
+
+
+# workaround for BinaryView.add_tag not supporting auto tags
+def add_auto_data_tag(bv: binaryninja.BinaryView, addr: int, name: str, desc: str) -> None:
+    tag = binaryninja.core.BNCreateTag(bv.get_tag_type(name).handle, desc)
+    binaryninja.core.BNAddTag(bv.handle, tag, False)
+    binaryninja.core.BNAddAutoDataTag(bv.handle, addr, tag)
+
+
+# try to add a function tag to the widest function containing the address
+# if there are none, resort to a data tag instead
+def add_auto_tag(bv: binaryninja.BinaryView, addr: int, name: str, desc: str) -> None:
+    f = get_widest_func(bv, addr)
+    if f is None:
+        add_auto_data_tag(bv, addr, name, desc)
+    else:
+        f.add_tag(name, desc, addr=addr, auto=True)
+
+
+# workaround for there to be no way to get all address tags in the python API
+def get_tag_refs(bv: binaryninja.BinaryView, ty: str) -> List[binaryninja.core.BNTagReference]:
+    tag_type = bv.get_tag_type(ty)
+    count = binaryninja.core.BNGetAllTagReferencesOfTypeCount(bv.handle, tag_type.handle)
+    ref_ptr = binaryninja.core.BNGetAllTagReferencesOfType(
+        bv.handle, tag_type.handle, ctypes.c_ulong(count)
+    )
+    return [ref_ptr[i] for i in range(count)]
```

Comment:
```suggestion
    return ref_ptr[:count]
```
or `return list(ref_ptr[:count])`?

Reply:
`ref_ptr` is a ctypes pointer but it turns out slicing is supported anyways so I'll change that.